### ISSUE TITLE
Undocumented node-types.json

### ISSUE THREAD

**Issue by pontaoski** [tree-sitter/tree-sitter#1663]
I'm interested in writing codegen using node-types.json for Go, but I can't figure out what exactly the schema of the file is supposed to be in order to generate an API from it.

**Comment by patrickt** [comment]
Yeah, we should definitely document this. The closest we have is this Haskell code, which might unblock you: https://github.com/github/semantic/blob/master/semantic-ast/src/AST/Deserialize.hs

Basically, they represent one of three things: a sum type (nodes representing choices between other types, like the concept of rvalues in C), a product type (nodes that declare one or more children), or a leaf node (something with no children, like a bare keyword.)

Example of sum types: https://github.com/tree-sitter/tree-sitter-php/blob/master/src/node-types.json#L2-L71
Product types: https://github.com/tree-sitter/tree-sitter-php/blob/master/src/node-types.json#L407-L426
Leaf types: https://github.com/tree-sitter/tree-sitter-php/blob/master/src/node-types.json#L5031-L5034

**Comment by dcreager** [comment]
There's [some documentation](https://tree-sitter.github.io/tree-sitter/using-parsers#static-node-types) here describing the schema (it also has a link to the Haskell code @patrickt mentions).
